README.md
# TOT (The One Thing)

[![Build Status](https://github.com/alanvardy/tot/workflows/ci/badge.svg)](https://github.com/alanvardy/tot) [![codecov](https://codecov.io/gh/alanvardy/tot/branch/master/graph/badge.svg?token=9FBJK1SU0K)](https://codecov.io/gh/alanvardy/tot) [![Crates.io](https://img.shields.io/crates/v/tot.svg)](https://crates.io/crates/tot)

A tiny GUI for doing the next thing on your Todoist list by priority.

![TOT](tot.png)

Requires [tod](https://github.com/alanvardy/tod) to be installed and run at least once.

## Keyboard Shortcuts

- `c` Complete task
- `h` Hide project


CHANGELOG.md
# Changelog

## Unreleased

- Update dependencies

## 2023-02-03 v0.1.5

- Add keyboard shortcut c for complete task
- Add keyboard shortcut h for hide project

## 2023-01-29 v0.1.4

- Show a spinner and don't block the main process when completing a task

## 2023-01-07 v0.1.3

- Add a hide project button

## 2023-01-04 v0.1.2

- Don't show the complete button when there is no task to complete

## 2023-01-04 v0.1.1

- Can select a project from `tod` config
- Added this changelog file
- Added publish checklist
- Add CI
- Add image to README.md
- Update dependencies

## 2023-01-02 v0.1.0

- Initial release
- Displays the highest priority task in the `home` project.
- Can complete said task and get the next one for the `home` project


src/test.rs
#[cfg(test)]
pub mod helpers {
    use crate::config;
    use crate::config::Config;
    use crate::items::{DateInfo, Item};
    use std::collections::HashMap;

    pub fn item_fixture() -> Item {
        Item {
            id: String::from("222"),
            content: String::from("Get gifts for the twins"),
            checked: false,
            description: String::from(""),
            due: Some(DateInfo {
                date: String::from("2061-11-13"),
                is_recurring: false,
                timezone: Some(String::from("America/Los_Angeles")),
            }),
            priority: 3,
            is_deleted: false,
        }
    }

    pub fn config_fixture() -> Config {
        Config {
            token: String::from("alreadycreated"),
            projects: HashMap::new(),
            path: config::generate_path().unwrap(),
            next_id: None,
            last_version_check: None,
            timezone: Some(String::from("US/Pacific")),
        }
    }
}
#[cfg(test)]
pub mod responses {

    pub fn sync() -> String {
        String::from(
            "{
    \"creator_id\": \"2671355\",
    \"created_at\": \"2019-12-11T22:36:50.000000Z\",
    \"assignee_id\": \"2671362\",
    \"assigner_id\": \"2671355\",
    \"comment_count\": 10,
    \"is_completed\": false,
    \"content\": \"Buy Coffee\",
    \"description\": \"\",
    \"due\": {
        \"date\": \"2016-09-01\",
        \"is_recurring\": false,
        \"datetime\": \"2016-09-01T12:00:00.000000Z\",
        \"string\": \"tomorrow at 12\",
        \"timezone\": \"Europe/Moscow\"
    },
    \"id\": \"2995104339\",
    \"labels\": [\"Food\", \"Shopping\"],
    \"order\": 1,
    \"priority\": 1,
    \"project_id\": \"2203306141\",
    \"section_id\": \"7025\",
    \"parent_id\": \"2995104589\",
    \"url\": \"https://todoist.com/showTask?id=2995104339\"
}",
        )
    }
}


src/main.rs
#![cfg_attr(not(debug_assertions), windows_subsystem = "windows")] // hide console window on Windows in release

use core::time::Duration;
use eframe::egui;
use std::sync::mpsc;
use std::thread;

mod config;
mod items;
mod projects;
mod request;
mod test;
mod time;

use config::Config;

fn main() {
    // Log to stdout (if you run with `RUST_LOG=debug`).
    tracing_subscriber::fmt::init();

    let options = eframe::NativeOptions {
        initial_window_size: Some(egui::vec2(320.0, 240.0)),
        ..Default::default()
    };
    eframe::run_native(
        "The One Thing",
        options,
        Box::new(|_cc| Box::<MyApp>::default()),
    )
    .unwrap();
}

#[derive(Clone)]
enum State {
    // Clear the screen and prepare to fetch
    BeginFetch,
    // Spawn a thread and fetch in the background
    Fetching,
    // Fetch is complete, show the results
    DoneFetch { text: Option<String> },
}

struct MyApp {
    projects: Vec<String>,
    project: String,
    state: State,
    tx: mpsc::Sender<Option<String>>,
    rx: mpsc::Receiver<Option<String>>,
}

impl Default for MyApp {
    fn default() -> Self {
        let projects = projects();
        let project = get_first_project(projects.clone());
        let text = get_next(project.clone());
        let (tx, rx) = mpsc::channel();

        Self {
            state: State::DoneFetch { text },
            projects,
            project,
            tx,
            rx,
        }
    }
}

#[allow(clippy::collapsible_else_if)]
#[allow(clippy::collapsible_if)]
impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        egui::CentralPanel::default().show(ctx, |ui| {
            ui.with_layout(egui::Layout::top_down(egui::Align::Center), |ui| {
                ui.label("The One Thing");
                ui.label(String::new());

                ui.vertical_centered(|ui| match self.state.clone() {
                    State::BeginFetch => {
                        ui.add(egui::Spinner::new());
                        spawn_complete_task(self.project.clone(), self.tx.clone());
                        self.state = State::Fetching;
                    }

                    State::Fetching => {
                        ui.add(egui::Spinner::new());

                        if let Ok(text) = self.rx.try_recv() {
                            self.state = State::DoneFetch { text };
                        }
                    }

                    State::DoneFetch { text } => {
                        if let Some(text) = text {
                            ui.heading(text);
                            ui.label(String::new());
                            if ui.button("Complete ✔").clicked() {
                                self.state = State::BeginFetch;
                            }
                            ui.label(String::new());
                            if ui.input(|i| i.key_pressed(egui::Key::C)) {
                                self.state = State::BeginFetch;
                            }
                        } else {
                            ui.heading(String::from("\nNo tasks remaining"));
                            ui.label(String::new());
                            ui.label(String::new());
                        };
                        if ui.button("Hide Project 🗙").clicked() {
                            hide(self.project.clone(), self);
                        }
                        if ui.input(|i| i.key_pressed(egui::Key::H)) {
                            hide(self.project.clone(), self);
                        }
                    }
                });
            });

            ui.with_layout(egui::Layout::left_to_right(egui::Align::BOTTOM), |ui| {
                for project in self.projects.iter() {
                    if *project.clone() == self.project {
                        if ui.add_enabled(false, egui::Button::new(project)).clicked() {
                            unreachable!();
                        }
                    } else {
                        if ui.button(project).clicked() {
                            self.project = project.to_string();
                            self.state = State::DoneFetch {
                                text: get_next(self.project.clone()),
                            };
                        }
                    }
                }
            });
        });
        ctx.request_repaint_after(Duration::new(0, 100));
    }
}

fn projects() -> Vec<String> {
    match config::get_or_create(None) {
        Ok(Config { projects, .. }) => {
            let mut projects = projects
                .keys()
                .map(|k| k.to_owned())
                .collect::<Vec<String>>();

            projects.sort();
            projects
        }
        Err(e) => vec![e],
    }
}

fn get_next(project: String) -> Option<String> {
    match config::get_or_create(None) {
        Ok(config) => projects::next_item(config, &project).unwrap(),
        Err(_e) => None,
    }
}

fn spawn_complete_task(project: String, tx: mpsc::Sender<Option<String>>) {
    thread::spawn(|| complete(project, tx));
}

fn complete(project: String, tx: mpsc::Sender<Option<String>>) {
    match config::get_or_create(None) {
        Ok(config) => {
            request::complete_item(config).unwrap();
            match get_next(project) {
                Some(text) => tx.send(Some(text)),
                None => tx.send(None),
            }
        }
        Err(_e) => tx.send(None),
    }
    .unwrap();
}

fn hide(project: String, state: &mut MyApp) {
    let projects: Vec<String> = state
        .projects
        .clone()
        .into_iter()
        .filter(|s| s != &project)
        .collect();

    let project = get_first_project(projects.clone());

    state.projects = projects;
    state.project = project.clone();
    state.state = State::DoneFetch {
        text: get_next(project),
    };
}

fn get_first_project(projects: Vec<String>) -> String {
    projects
        .first()
        .map(|f| f.to_string())
        .unwrap_or_else(|| String::from("No projects found"))
}


src/config.rs
use crate::time;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::collections::HashMap;
use std::io::{Read, Write};
use std::{fs, io};

/// App configuration, serialized as json in $XDG_CONFIG_HOME/tod.cfg
#[derive(Clone, Serialize, Deserialize, Eq, PartialEq, Debug)]
pub struct Config {
    /// The Todoist Api token
    pub token: String,
    /// List of Todoist projects and their project numbers
    pub projects: HashMap<String, u32>,
    /// Path to config file
    pub path: String,
    /// The ID of the next task
    pub next_id: Option<String>,
    pub timezone: Option<String>,
    pub last_version_check: Option<String>,
}

impl Config {
    pub fn new(token: &str) -> Result<Config, String> {
        let projects: HashMap<String, u32> = HashMap::new();
        Ok(Config {
            path: generate_path()?,
            token: String::from(token),
            next_id: None,
            last_version_check: None,
            timezone: None,
            projects,
        })
    }

    pub fn create(self) -> Result<Config, String> {
        let json = json!(self).to_string();
        let mut file = fs::File::create(&self.path).or(Err("Could not create file"))?;
        file.write_all(json.as_bytes())
            .or(Err("Could not write to file"))?;
        println!("Config successfully created in {}", &self.path);
        Ok(self)
    }

    pub fn save(self) -> std::result::Result<String, String> {
        let json = json!(self);
        let string = serde_json::to_string_pretty(&json).or(Err("Could not convert to JSON"))?;

        fs::OpenOptions::new()
            .write(true)
            .read(true)
            .truncate(true)
            .open(&self.path)
            .or(Err("Could not find config"))?
            .write_all(string.as_bytes())
            .or(Err("Could not write to file"))?;

        Ok(String::from("✓"))
    }

    pub fn load(path: &str) -> Result<Config, String> {
        let mut json = String::new();

        fs::File::open(path)
            .or(Err("Could not find file"))?
            .read_to_string(&mut json)
            .or(Err("Could not read to string"))?;

        serde_json::from_str::<Config>(&json).map_err(|_| String::from("Could not parse JSON"))
    }

    pub fn set_path(self, path: &str) -> Config {
        Config {
            path: String::from(path),
            ..self
        }
    }

    pub fn set_next_id(&self, next_id: String) -> Config {
        let next_id: Option<String> = Some(next_id);

        Config {
            next_id,
            ..self.clone()
        }
    }

    pub fn clear_next_id(self) -> Config {
        let next_id: Option<String> = None;

        Config { next_id, ..self }
    }

    fn check_for_timezone(self: Config) -> Result<Config, String> {
        if self.timezone.is_none() {
            time::list_timezones();
            let desc = "Please enter the number of your timezone";
            let num: usize = get_input(desc)?
                .parse::<usize>()
                .map_err(|_| String::from("Could not parse string into number"))?;
            let config = Config {
                timezone: Some(time::get_timezone(num)),
                ..self
            };

            config.clone().save()?;

            Ok(config)
        } else {
            Ok(self)
        }
    }
}

pub fn get_or_create(config_path: Option<&str>) -> Result<Config, String> {
    let path: String = match config_path {
        None => generate_path()?,
        Some(path) => String::from(path).trim().to_owned(),
    };
    let desc = "Please enter your Todoist API token from https://todoist.com/prefs/integrations ";

    if !path_exists(&path) {
        // We used to store config in $HOME/.tod.cfg
        // This moves it to new path
        let legacy_path = generate_legacy_path()?;
        if path_exists(&legacy_path) {
            println!("INFO: Moving the config file from \"{legacy_path}\" to \"{path}\".\n");
            fs::rename(legacy_path, &path).map_err(|e| e.to_string())?;
        }
    }

    match fs::File::open(&path) {
        Ok(_) => {
            let config = Config::load(&path)?;

            // When we move the config file we also need to rename the path in JSON
            if config.path != path {
                let new_config = config.set_path(&path);
                new_config.clone().save()?;
                Ok(new_config)
            } else {
                Ok(config)
            }
        }
        Err(_) => {
            let token = get_input(desc)?;
            Config::new(&token)?.create()?.check_for_timezone()
        }
    }
}

fn path_exists(path: &str) -> bool {
    std::path::Path::new(path).exists()
}

pub fn generate_path() -> Result<String, String> {
    let filename = if cfg!(test) { "test" } else { "tod.cfg" };

    let config_directory = dirs::config_dir()
        .ok_or_else(|| String::from("Could not find config directory"))?
        .to_str()
        .ok_or_else(|| String::from("Could not convert config directory to string"))?
        .to_owned();
    Ok(format!("{config_directory}/{filename}"))
}

pub fn generate_legacy_path() -> Result<String, String> {
    let filename = if cfg!(test) { "test" } else { ".tod.cfg" };

    let home_directory = dirs::home_dir()
        .ok_or_else(|| String::from("Could not find home directory"))?
        .to_str()
        .ok_or_else(|| String::from("Could not convert directory to string"))?
        .to_owned();
    Ok(format!("{home_directory}/{filename}"))
}

pub fn get_input(desc: &str) -> Result<String, String> {
    if cfg!(test) {
        return Ok(String::from("5"));
    }

    let mut input = String::new();
    println!("{desc}");
    io::stdin()
        .read_line(&mut input)
        .or(Err("error: unable to read user input"))?;

    Ok(String::from(input.trim()))
}


src/request.rs
use reqwest::blocking::Client;
use reqwest::header::AUTHORIZATION;
use reqwest::header::CONTENT_TYPE;
use serde_json::json;
use uuid::Uuid;

use crate::config::Config;
use crate::items;
use crate::items::Item;

#[cfg(test)]
use mockito;

// TODOIST URLS
const PROJECT_DATA_URL: &str = "/sync/v9/projects/get_data";
const SYNC_URL: &str = "/sync/v9/sync";

const FAKE_UUID: &str = "42963283-2bab-4b1f-bad2-278ef2b6ba2c";

/// Get a vector of all items for a project
pub fn items_for_project(config: &Config, project_id: &str) -> Result<Vec<Item>, String> {
    let url = String::from(PROJECT_DATA_URL);
    let body = json!({ "project_id": project_id });
    let json = post_todoist_sync(config.token.clone(), url, body)?;
    items::json_to_items(json)
}

/// Complete the last item returned by "next item"
pub fn complete_item(config: Config) -> Result<String, String> {
    let body = json!({"commands": [{"type": "item_close", "uuid": new_uuid(), "temp_id": new_uuid(), "args": {"id": config.next_id}}]});
    let url = String::from(SYNC_URL);

    post_todoist_sync(config.token.clone(), url, body)?;

    if !cfg!(test) {
        config.clear_next_id().save()?;
    }

    // Does not pass back an item
    Ok(String::from("✓"))
}

/// Post to Todoist via sync API
fn post_todoist_sync(
    token: String,
    url: String,
    body: serde_json::Value,
) -> Result<String, String> {
    #[cfg(not(test))]
    let todoist_url: &str = "https://api.todoist.com";

    #[cfg(test)]
    let todoist_url: &str = &mockito::server_url();

    let request_url = format!("{todoist_url}{url}");

    let response = Client::new()
        .post(request_url)
        .header(CONTENT_TYPE, "application/json")
        .header(AUTHORIZATION, format!("Bearer {token}"))
        .json(&body)
        .send()
        .or(Err("Did not get response from server"))?;

    if response.status().is_success() {
        Ok(response.text().or(Err("Could not read response text"))?)
    } else {
        Err(format!("Error: {:#?}", response.text()))
    }
}

/// Create a new UUID, required for Todoist API
fn new_uuid() -> String {
    if cfg!(test) {
        String::from(FAKE_UUID)
    } else {
        Uuid::new_v4().to_string()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test;
    use pretty_assertions::assert_eq;

    #[test]
    fn should_complete_an_item() {
        let _m = mockito::mock("POST", "/sync/v9/sync")
            .with_status(200)
            .with_header("content-type", "application/json")
            .with_body(&test::responses::sync())
            .create();

        let config = Config::new("12341234")
            .unwrap()
            .set_next_id(String::from("112233"));
        let response = complete_item(config);
        assert_eq!(response, Ok(String::from("✓")));
    }
}


src/projects.rs
use crate::config::Config;
use crate::{items, projects, request};

pub fn project_id(config: &Config, project_name: &str) -> Result<String, String> {
    let project_id = config
        .projects
        .get(project_name)
        .ok_or(format!(
            "Project {project_name} not found, please add it to config"
        ))?
        .to_string();

    Ok(project_id)
}

/// Get the next item by priority and save its id to config
pub fn next_item(config: Config, project_name: &str) -> Result<Option<String>, String> {
    let project_id = projects::project_id(&config, project_name)?;
    let items = request::items_for_project(&config, &project_id)?;
    let filtered_items = items::filter_not_in_future(items, &config)?;
    let maybe_item = items::sort_by_value(filtered_items, &config)
        .first()
        .map(|item| item.to_owned());

    match maybe_item {
        Some(item) => {
            config.set_next_id(item.id.clone()).save()?;
            Ok(Some(item.fmt(&config)))
        }
        None => Ok(None),
    }
}


src/items.rs
use chrono::DateTime;
use chrono::NaiveDate;
use chrono_tz::Tz;
use serde::{Deserialize, Serialize};
use std::cmp::Reverse;

use crate::config::Config;
use crate::time;

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct Item {
    pub id: String,
    pub content: String,
    pub priority: u8,
    pub checked: bool,
    pub description: String,
    pub due: Option<DateInfo>,
    pub is_deleted: bool,
}

#[derive(Serialize, Deserialize, Debug, Clone, Eq, PartialEq)]
pub struct DateInfo {
    pub date: String,
    pub is_recurring: bool,
    pub timezone: Option<String>,
}

#[derive(Serialize, Deserialize, Debug)]
struct Body {
    items: Vec<Item>,
}

enum DateTimeInfo {
    NoDateTime,
    Date {
        date: NaiveDate,
        is_recurring: bool,
    },
    DateTime {
        datetime: DateTime<Tz>,
        is_recurring: bool,
    },
}

impl Item {
    pub fn fmt(&self, config: &Config) -> String {
        let description = match &*self.description {
            "" => String::from(""),
            _ => format!("\n{}", self.description),
        };

        let due = match &self.datetimeinfo(config) {
            Ok(DateTimeInfo::Date { date, is_recurring }) => {
                let recurring_icon = if *is_recurring { " ↻" } else { "" };
                let date_string = time::format_date(date, config);

                format!("\nDue: {date_string}{recurring_icon}")
            }
            Ok(DateTimeInfo::DateTime {
                datetime,
                is_recurring,
            }) => {
                let recurring_icon = if *is_recurring { " ↻" } else { "" };
                let datetime_string = time::format_datetime(datetime, config);

                format!("\nDue: {datetime_string}{recurring_icon}")
            }
            Ok(DateTimeInfo::NoDateTime) => String::from(""),
            Err(string) => string.clone(),
        };

        format!("\n{}{}{}", self.content, description, due)
    }

    /// Determines the numeric value of an item for sorting
    fn value(&self, config: &Config) -> u32 {
        let date_value: u8 = self.date_value(config);
        let priority_value: u8 = self.priority_value();

        date_value as u32 + priority_value as u32
    }

    /// Return the value of the due field
    fn date_value(&self, config: &Config) -> u8 {
        match &self.datetimeinfo(config) {
            Ok(DateTimeInfo::NoDateTime) => 80,
            Ok(DateTimeInfo::Date { date, is_recurring }) => {
                let today_value = if *date == time::today_date(config) {
                    100
                } else {
                    0
                };
                let overdue_value = if self.is_overdue(config) { 150 } else { 0 };
                let recurring_value = if is_recurring.to_owned() { 0 } else { 50 };
                today_value + overdue_value + recurring_value
            }
            Ok(DateTimeInfo::DateTime {
                datetime,
                is_recurring,
            }) => {
                let recurring_value = if is_recurring.to_owned() { 0 } else { 50 };
                let duration = *datetime - time::now(config);
                match duration.num_minutes() {
                    -15..=15 => 200 + recurring_value,
                    _ => recurring_value,
                }
            }
            Err(_) => 50,
        }
    }

    fn priority_value(&self) -> u8 {
        match self.priority {
            2 => 1,
            3 => 3,
            4 => 4,
            _ => 2,
        }
    }

    /// Converts the JSON date representation into Date or Datetime
    fn datetimeinfo(&self, config: &Config) -> Result<DateTimeInfo, String> {
        let tz = match (self.clone().due, config.clone().timezone) {
            (None, Some(tz_string)) => time::timezone_from_str(&Some(tz_string)),
            (None, None) => Tz::UTC,
            (Some(DateInfo { timezone: None, .. }), Some(tz_string)) => time::timezone_from_str(&Some(tz_string)),
            (Some(DateInfo { timezone: None, .. }), None) => Tz::UTC,
            (Some(DateInfo {
                timezone: Some(tz_string),
                ..
                // Remove the Some here
            }), _) => time::timezone_from_str(&Some(tz_string)),
        };
        match self.clone().due {
            None => Ok(DateTimeInfo::NoDateTime),
            Some(DateInfo {
                date, is_recurring, ..
            }) if date.len() == 10 => Ok(DateTimeInfo::Date {
                date: time::date_from_str(&date, tz)?,
                is_recurring,
            }),
            Some(DateInfo {
                date, is_recurring, ..
            }) => Ok(DateTimeInfo::DateTime {
                datetime: time::datetime_from_str(&date, tz)?,
                is_recurring,
            }),
        }
    }

    fn has_no_date(&self) -> bool {
        self.due.is_none()
    }

    // Returns true if the datetime is today and there is a time
    fn is_today(&self, config: &Config) -> bool {
        match self.datetimeinfo(config) {
            Ok(DateTimeInfo::NoDateTime) => false,
            Ok(DateTimeInfo::Date { date, .. }) => date == time::today_date(config),
            Ok(DateTimeInfo::DateTime { datetime, .. }) => {
                time::datetime_is_today(datetime, config)
            }
            Err(_) => false,
        }
    }

    fn is_overdue(&self, config: &Config) -> bool {
        match self.clone().datetimeinfo(config) {
            Ok(DateTimeInfo::NoDateTime) => false,
            Ok(DateTimeInfo::Date { date, .. }) => time::is_date_in_past(date, config),
            Ok(DateTimeInfo::DateTime { datetime, .. }) => {
                time::is_date_in_past(datetime.date_naive(), config)
            }
            Err(_) => false,
        }
    }
}
pub fn json_to_items(json: String) -> Result<Vec<Item>, String> {
    let result: Result<Body, _> = serde_json::from_str(&json);
    match result {
        Ok(body) => Ok(body.items),
        Err(err) => Err(format!("Could not parse response for item: {err:?}")),
    }
}

pub fn sort_by_value(mut items: Vec<Item>, config: &Config) -> Vec<Item> {
    items.sort_by_key(|b| Reverse(b.value(config)));
    items
}

pub fn filter_not_in_future(items: Vec<Item>, config: &Config) -> Result<Vec<Item>, String> {
    let items = items
        .into_iter()
        .filter(|item| item.is_today(config) || item.has_no_date() || item.is_overdue(config))
        .collect();

    Ok(items)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test;
    use pretty_assertions::assert_eq;

    #[test]
    fn date_value_can_handle_date() {
        let config = test::helpers::config_fixture();
        // On another day
        assert_eq!(test::helpers::item_fixture().date_value(&config), 50);

        // Recurring
        let item = Item {
            due: Some(DateInfo {
                is_recurring: true,
                ..test::helpers::item_fixture().due.unwrap()
            }),
            ..test::helpers::item_fixture()
        };
        assert_eq!(item.date_value(&config), 0);

        // Overdue
        let item = Item {
            due: Some(DateInfo {
                date: String::from("2001-11-13"),
                is_recurring: true,
                timezone: Some(String::from("America/Los_Angeles")),
            }),
            ..test::helpers::item_fixture()
        };
        assert_eq!(item.date_value(&config), 150);

        // No date
        let item = Item { due: None, ..item };
        assert_eq!(item.date_value(&config), 80);
    }

    #[test]
    fn date_value_can_handle_datetime() {
        let config = test::helpers::config_fixture();
        let item = Item {
            due: Some(DateInfo {
                date: String::from("2021-02-27T19:41:56Z"),
                ..test::helpers::item_fixture().due.unwrap()
            }),
            ..test::helpers::item_fixture()
        };

        assert_eq!(item.date_value(&config), 50);
    }

    #[test]
    fn can_format_item_with_a_date() {
        let config = test::helpers::config_fixture();
        let item = Item {
            content: String::from("Get gifts for the twins"),
            due: Some(DateInfo {
                date: String::from("2021-08-13"),
                ..test::helpers::item_fixture().due.unwrap()
            }),
            ..test::helpers::item_fixture()
        };

        let output = "\nGet gifts for the twins\nDue: 2021-08-13";

        assert_eq!(format!("{}", item.fmt(&config)), output);
    }

    #[test]
    fn can_format_item_with_today() {
        let config = test::helpers::config_fixture();
        let item = Item {
            content: String::from("Get gifts for the twins"),
            due: Some(DateInfo {
                date: time::today_string(&config),
                ..test::helpers::item_fixture().due.unwrap()
            }),
            ..test::helpers::item_fixture()
        };

        let output = "\nGet gifts for the twins\nDue: Today";

        assert_eq!(format!("{}", item.fmt(&config)), output);
    }

    #[test]
    fn value_can_get_the_value_of_an_item() {
        let config = test::helpers::config_fixture();
        let item = test::helpers::item_fixture();

        assert_eq!(item.value(&config), 53);
    }

    #[test]
    fn has_no_date_works() {
        let config = test::helpers::config_fixture();
        let item = Item {
            due: None,
            ..test::helpers::item_fixture()
        };

        assert!(item.has_no_date());

        let item_today = Item {
            due: Some(DateInfo {
                date: time::today_string(&config),
                ..test::helpers::item_fixture().due.unwrap()
            }),
            ..test::helpers::item_fixture()
        };
        assert!(!item_today.has_no_date());
    }

    #[test]
    fn is_today_works() {
        let config = test::helpers::config_fixture();
        let item = Item {
            due: None,
            ..test::helpers::item_fixture()
        };

        assert!(!item.is_today(&config));

        let item_today = Item {
            due: Some(DateInfo {
                date: time::today_string(&config),
                is_recurring: false,
                timezone: None,
            }),
            ..test::helpers::item_fixture()
        };
        assert!(item_today.is_today(&config));

        let item_in_past = Item {
            due: Some(DateInfo {
                date: String::from("2021-09-06T16:00:00"),
                is_recurring: false,
                timezone: None,
            }),
            ..test::helpers::item_fixture()
        };
        assert!(!item_in_past.is_today(&config));
    }

    #[test]
    fn sort_by_value_works() {
        let config = test::helpers::config_fixture();
        let today = Item {
            due: Some(DateInfo {
                date: time::today_string(&config),
                is_recurring: false,
                timezone: None,
            }),
            ..test::helpers::item_fixture()
        };

        let today_recurring = Item {
            due: Some(DateInfo {
                date: time::today_string(&config),
                is_recurring: false,
                timezone: None,
            }),
            ..test::helpers::item_fixture()
        };

        let future = Item {
            due: Some(DateInfo {
                date: String::from("2035-12-12"),
                is_recurring: false,
                timezone: None,
            }),
            ..test::helpers::item_fixture()
        };

        let input = vec![future.clone(), today_recurring.clone(), today.clone()];
        let result = vec![today, today_recurring, future];

        assert_eq!(sort_by_value(input, &config), result);
    }

    #[test]
    fn is_overdue_works() {
        let config = test::helpers::config_fixture();
        let item = Item {
            id: String::from("222"),
            content: String::from("Get gifts for the twins"),
            checked: false,
            description: String::from(""),
            due: None,
            priority: 3,
            is_deleted: false,
        };

        assert!(!item.is_overdue(&config));

        let item_today = Item {
            due: Some(DateInfo {
                date: time::today_string(&config),
                is_recurring: false,
                timezone: None,
            }),
            ..item.clone()
        };
        assert!(!item_today.is_overdue(&config));

        let item_future = Item {
            due: Some(DateInfo {
                date: String::from("2035-12-12"),
                is_recurring: false,
                timezone: None,
            }),
            ..item.clone()
        };
        assert!(!item_future.is_overdue(&config));

        let item_today = Item {
            due: Some(DateInfo {
                date: String::from("2020-12-20"),
                is_recurring: false,
                timezone: None,
            }),
            ..item
        };
        assert!(item_today.is_overdue(&config));
    }

    #[test]
    fn json_to_items_works() {
        let json = String::from("2{.e");
        let error_text = String::from("Could not parse response for item: Error(\"invalid type: integer `2`, expected struct Body\", line: 1, column: 1)");
        assert_eq!(json_to_items(json), Err(error_text));
    }
}
